apple/swift-docc
# Validate and filter parameter and return value documentation

Bug/issue #, if applicable: rdar://118739612

## Summary

This adds validation and filtering of parameter and return value documentation based on a symbol's function signature in each language representation.

This results in 3 new behaviors:
- A parameter or return value that's not applicable to a specific language representation of a symbol isn't displayed in that language's variant of the page.
- Objective-C error parameters or return values that are not documented for symbols defined in other languages are synthesized if any other parameters are documented.
- User facing diagnostics with fixits are emitted for parameters that are misspelled, not found, or repeated and for return values for symbols that return void in all language representations.

### Filtering parameters

If a symbol has different parameters and return values in different languages, then each language's version of that symbol's page will only display the parameters that's applicable to that language. 

This is for example relevant for API with errors in Objective-C that bridge to API that `throws` in Swift;

```objc
/// Does something in Objective-C
/// - Parameters:
///   - someValue: Some value.
///   - error: On input, a pointer to an error object. If an error occurs, this pointer is set to an actual error object containing the error information.
/// - Returns: `YES` if doing something was successful, or `NO` if an error occurred.
- (BOOL)doSomethingWith:(NSInteger)someValue
                  error:(NSError **)error;
```

This method become `func doSomething(with someValue: Int) throws` in Swift which returns `Void` and doesn't have an `error` parameter, so displaying documentation for the error parameter and the return value wouldn't be helpful for someone reading the Swift version of that method's documentation. 

Before this PR we displayed all documented parameters in all language variants of a symbol's page. With these changes, the Swift page won't display the "error" parameter documentation or the return value documentation.

### Adding Objective-C error documentation

The above works well for functions defined in Objective-C that document parameters and return values that aren't available in Swift, but it doesn't work in the other direction;

```swift
/// Does something in Swift.
/// - Parameter someValue: Some value.
/// - Throws: Some error if something does wrong
@objc public func doSomething(with someValue: Int) throws { }
```

This becomes `- (BOOL)doSomethingWith:(NSInteger)someValue error:(NSError **)error;` in Objective-C but the Swift documentation comment doesn't have any documentation for the Objective-C "error" parameter or the Objective-C specific return value.

Before this PR we only displayed the developer documented parameters. With these changes, the Objective-C page will display a synthesized "error" parameter documentation with the generic description:

> **error**
> On input, a pointer to an error object. If an error occurs, this pointer is set to an actual error object containing the error information. 

and a synthesized return value documentation with the generic description:
> **Return Value**
>`YES` if successful, or `NO` if an error occurred.

If the throwing Swift function was defined with a non-null return value that bridges to a nullable return value in Objective-C;

```swift
/// Returns something from Swift.
/// - Returns: Some string.
/// - Throws: Some error if something does wrong
@objc public func returnSomething() throws -> String { "" }
```

Then the Objective-C return value documentation is appended with a generic description of this automatic bridging behavior:

> **Return Value**
> Some string. If an error occurs, this method returns `nil` and assigns an appropriate error object to the `error` parameter.

### Parameter and return value diagnostics 

If the symbol documents at least one parameter or its return value, then DocC will validate that symbol's parameters section and return value section. Specifically, DocC will warn in 5 cases:
1. A parameter is documented more than once
2. A parameter that doesn't exist in the function signature is documented
3. A parameter from the function signature isn't documented
4. A parameter is documented using its argument label instead of its parameter name
5. A symbol that returns `Void` has documented a return value

The Swift function below shows minimal examples of each of these cases:

```swift
/// - Parameters:
///   - first: Something
///   - first: Something else (1)
///   - with: Something (4)
///   - third: Something (2)
///  (3)
/// - Returns: Something (5)
func doSomething(with first: Int, and second: Int) { }
```

## Dependencies

https://github.com/apple/swift-docc-symbolkit/pull/64

## Testing

- In a multi-language project, add:
  - Objective-C API with errors with documentation for all parameters and return values
  - Swift `@objc` API that `throws` with documentation for the Swift parameters.

- Build documentation for the project and pass `--enable-experimental-parameters-and-returns-validation`.

- View both the Swift and Objective-C versions of the documentation pages for both the API defined in Objective-C and the API defined in Swift. 
  - Each language variant of the documentation pages should only display the parameters and return values applicable to that language. 

- Add parameter documentation for parameters that doesn't exist to the Swift and Objective-C definitions and build documentation again. 

- View both the Swift and Objective-C versions of the documentation pages for both the API defined in Objective-C and the API defined in Swift. 
  - The documentation for the parameters that doesn't exist shouldn't be displayed on either languages's variant of the documentation page.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
